src/ln/channelmonitor.rs

```diff
@@ -1471,107 +1475,115 @@ impl ChannelMonitor {
-					} else { None };
-				if let Some(revocation_point) = revocation_point_option {
-					let (revocation_pubkey, b_htlc_key) = match self.key_storage {
-						Storage::Local { ref revocation_base_key, ref htlc_base_key, .. } => {
-							(ignore_error!(chan_utils::derive_public_revocation_key(&self.secp_ctx, revocation_point, &PublicKey::from_secret_key(&self.secp_ctx, &revocation_base_key))),
-							ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, revocation_point, &PublicKey::from_secret_key(&self.secp_ctx, &htlc_base_key))))
-						},
-						Storage::Watchtower { ref revocation_base_key, ref htlc_base_key, .. } => {
-							(ignore_error!(chan_utils::derive_public_revocation_key(&self.secp_ctx, revocation_point, &revocation_base_key)),
-							ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, revocation_point, &htlc_base_key)))
-						},
-					};
-					let a_htlc_key = match self.their_htlc_base_key {
-						None => return (txn_to_broadcast, (commitment_txid, watch_outputs), spendable_outputs),
-						Some(their_htlc_base_key) => ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, revocation_point, &their_htlc_base_key)),
-					};
+				log_trace!(self, "Got broadcast of non-revoked remote commitment transaction {}", commitment_txid);
 
-					for (idx, outp) in tx.output.iter().enumerate() {
-						if outp.script_pubkey.is_v0_p2wpkh() {
-							match self.key_storage {
-								Storage::Local { ref payment_base_key, .. } => {
-									if let Ok(local_key) = chan_utils::derive_private_key(&self.secp_ctx, &revocation_point, &payment_base_key) {
+				let (revocation_pubkey, b_htlc_key) = match self.key_storage {
+					Storage::Local { ref revocation_base_key, ref htlc_base_key, .. } => {
+						(ignore_error!(chan_utils::derive_public_revocation_key(&self.secp_ctx, revocation_point, &PublicKey::from_secret_key(&self.secp_ctx, &revocation_base_key))),
+						ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, revocation_point, &PublicKey::from_secret_key(&self.secp_ctx, &htlc_base_key))))
+					},
+					Storage::Watchtower { ref revocation_base_key, ref htlc_base_key, .. } => {
+						(ignore_error!(chan_utils::derive_public_revocation_key(&self.secp_ctx, revocation_point, &revocation_base_key)),
+						ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, revocation_point, &htlc_base_key)))
+					},
+				};
+				let a_htlc_key = match self.their_htlc_base_key {
+					None => return (txn_to_broadcast, (commitment_txid, watch_outputs), spendable_outputs),
+					Some(their_htlc_base_key) => ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, revocation_point, &their_htlc_base_key)),
+				};
+
+				for (idx, outp) in tx.output.iter().enumerate() {
+					if outp.script_pubkey.is_v0_p2wpkh() {
+						match self.key_storage {
+							Storage::Local { ref payment_base_key, .. } => {
+								if let Ok(local_key) = chan_utils::derive_private_key(&self.secp_ctx, &revocation_point, &payment_base_key) {
```

Comment:
I dont think we want to do crypto operations in a loop over every output in every tx in a block that matched. Not entirely your fault, but this used to be in the match for it being definitely a commitment tx.

Reply:
Yes I know, that's why I've added another filter in block_connected based on nSequence. I've improved it, it should only match commitment txn with the lightning pattern (0x20 in nlocktime, 0x80 in nSequence). Could this be a DoS vector for watchtower ? Can we do better than just saying "we want to match this commitment transaction with the given output" ? Cache the script?

Reply:
Ok cached the script and strained filters.